File: Python/primsalgorithm.py

import sys
class Graph:
    def __init__(self,v):
        self.vertices = v
        self.graph = [[0 for i in range(v)] for i in range(v)]

    def add_weighted_undirected_edge(self,u,v,w):
        self.graph[u][v] = w
        self.graph[v][u] = w

    def minweight(self,key,mst):
        min = 100000000000
        minindex = -1
        for i in range(self.vertices):
            if key[i] < min and mst[i] == False:
                min = key[i]
                minindex = i
        return minindex

    def prims(self):
        mst = [False]*self.vertices
        keys = [100000000000]*(self.vertices)
        keys[0] = 0
        parent = [0]*self.vertices
        parent[0] = -1  
        for cout in range(self.vertices):
            u = self.minweight(keys, mst)
            mst[u] = True
            for v in range(self.vertices):
                if self.graph[u][v] > 0 and mst[v] == False and keys[v] > self.graph[u][v]:
                        keys[v] = self.graph[u][v]
                        parent[v] = u
        print("Mst is :")
        for i in range(self.vertices):
            print("{}-->{} -- {}".format(parent[i],i,self.graph[i][parent[i]]))
        print()

    def print_graph(self):
        for i in self.graph:
            print(i)

x = Graph(5)
x.print_graph()
while True:
    u,v,w = input().split()
    if (u!='n' or v!='n' or w!='n'):
        x.add_weighted_undirected_edge(int(u),int(v),int(w))
    else:
        break

x.print_graph()
x.prims()

File: C/PrimsAlgorithm.c
#include <stdio.h>

void print_graph(int n,int graph[100][100])
{
    int i,j;
    for(i=0;i<n;i++)
    {
        for(j=0;j<n;j++)
        {
            printf("%d ",graph[i][j]);
        }
        printf("\n");
    }
}



void create_graph(int n, int graph[100][100])
{
    int i,j;
    for(i=0;i<n;i++)
    {
        for(j=0;j<n;j++)
        {
            graph[i][j] = 0;
        }
    }
}

void enter_edge(int graph[100][100],int a,int b)
{
    graph[a][b] = 1;
    graph[b][a] = 1;
}

void enter_directed_edge(int graph[100][100],int a,int b)
{
    graph[a][b] = 1;
}

void enter_weighted_undirected_edge(int graph[100][100],int a,int b,int w)
{
    graph[a][b] = w;
    graph[b][a] = w;
}

void enter_directed_weighted_edge(int graph[100][100],int a,int b,int w)
{
    graph[a][b] = w;
}

int minweight(int key[100],int mst[100],int n)
{
    int min  = 100000000;
    int minindex = -1;
    for(int i =0;i<n;i++)
    {
        if(key[i]< min && mst[i] == 0)
        {
            min = key[i];
            minindex = i;
        }
    }
    return minindex;
}

void PrimsMST(int graph[100][100],int n)
{
   int mst[100],keys[100],parent[110];
   for(int t = 0;t<n;t++)
   {
       keys[t] = 100000000;
   }
   int u;
   parent[0] = -1;
   for(int k=0;k<n;k++)
   {
       u = minweight(keys,mst,n);
       mst[u] = 1;
       for(int v = 0;v<n;v++)
       {
           if(graph[u][v]>0 && mst[v]==0 && keys[v]>graph[u][v])
           {
               keys[v] = graph[u][v];
               parent[v] = u;
           }
       }
   }
   for(int i = 0;i<n;i++)
   {
       printf("%d --> %d -- %d \n",parent[i],i,graph[i][parent[i]]);
   }
   printf("\n"); 
}


int main()
{
    int n, graph[100][100];
    int u,v,w;
    printf("enter number of vertices \n");
    scanf("%d",&n);
    printf("adjacency matrix \n");
    create_graph(n,graph);
    print_graph(n,graph);
    while(1)
    {
        printf("Enter two vertices and weight(non zero) to create a directed weighted edge (enter 10000 10000 10000to exit) \n");
        scanf("%d %d %d",&u,&v,&w);
        if(u==10000 && v == 10000)
        {
            break;
        }
        enter_weighted_undirected_edge(graph,u,v,w);       
    }
    print_graph(n,graph);
    printf("MST(Prim's Algorithm) for the graph is \n");
    PrimsMST(graph,n);
    return 0;
}